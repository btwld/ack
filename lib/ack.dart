library;

export 'src/ack.dart';
