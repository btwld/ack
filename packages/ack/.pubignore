# Development and IDE files
melos_ack.iml
.claude/

# Build artifacts
build/
coverage/

# Local example (if any development examples exist)
example/
