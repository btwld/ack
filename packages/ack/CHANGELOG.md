## 0.2.0-beta.1 (2025-05-03)

* See [release notes](https://github.com/btwld/ack/releases/tag/v0.2.0-beta.1) for details.


## 0.2.0-beta.1

 - Bump "ack" to `0.2.0-beta.1`.

## 0.2.0

 - Bump "ack" to 0.2.0 with improved SchemaModel API and enhanced string validation

## 0.2.0

### Breaking Changes

- **SchemaModel API**: Completely redesigned the SchemaModel API for a more intuitive and streamlined experience
  - Changed constructor parameter from `Map<String, dynamic>` to `Object?` for better flexibility
  - Added automatic validation during construction
  - Added `isValid` getter and `getErrors()` method for easier error handling
  - Added abstract `getSchema()` method to define the schema
  - Removed `parse()`, `fromValidated()`, `validateMap()`, `validate()`, `createFromMap()`, and `initialize()` methods

### Improvements

- Improved error handling with more detailed error messages
- Better type safety with `Object?` instead of `dynamic`
- Simplified API with fewer methods and more intuitive usage
- **String Validation**: Enhanced string validation with improved methods
  - Changed `matches()` to validate full string patterns (anchored with ^ and $)
  - Added `contains()` method for partial string matching

## 0.1.2

- Improved error messages

## 0.1.1

-  Added deprecations

## 0.1.0

-  Error and exception improvements
-  Reworked validation and constraint workflows
-  Improved testing

## 0.0.2

-  Better JSON response parsing for OpenApiSchemaConverter

## 0.0.1

-  Initial version.