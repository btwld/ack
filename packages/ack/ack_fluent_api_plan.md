# Ack Fluent API Refactor Plan

This document outlines the plan to implement a fluent, Zod-inspired API for the Ack validation library. The work is broken down by schema type and is based on a detailed analysis of our current architecture and the capabilities of the Zod library.

---

### **Guiding Principles**

1.  **Universal Methods go on the Base Class**: Any capability that can apply to *any* schema (like making it nullable or adding a description) should be a method on the base `AckSchema`.
2.  **Specific Methods go in Extensions**: Capabilities that only make sense for a specific type (e.g., email validation for strings) should be implemented as extension methods on that specific schema class (`StringSchema`, `ListSchema`, etc.). This keeps the core classes clean.
3.  **Unsupported Features**: We will clearly mark features from Zod that we cannot or should not implement right now due to significant architectural differences (e.g., transformations, async validation).

---

### **1. `AckSchema` (Base Class Methods)**

These methods will be added directly to the `AckSchema<T>` class because they are universal.

| Method Signature | Zod Equivalent | Why Here? | Implementation Notes |
| :--- | :--- | :--- | :--- |
| `nullable()` | `.nullable()` | Makes any schema type nullable. This is the most fundamental and universal operation. | Already implemented via the `.nullable()` extension. No new work needed. |
| `default(T value)` | `.default(value)` | Sets a default value if the input is `null` or `undefined`. A universal concept. | Already implemented via the `defaultValue` constructor parameter. The `copyWith` pattern allows us to set it fluently. |
| `description(String text)` | `.describe(text)` | Adds a description for documentation/error message purposes. Universal. | Already implemented via the `description` constructor parameter. |

**Conclusion:** The most fundamental universal methods are already supported by our new architecture.

---

### **2. `StringSchema` Extensions**

These will be implemented in a new file: `lib/src/schemas/extensions/string_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `minLength(int n)` | `.min(n)` | String-specific length validation. | **Can Support Now** |
| `maxLength(int n)` | `.max(n)` | String-specific length validation. | **Can Support Now** |
| `length(int n)` | `.length(n)` | String-specific length validation. | **Can Support Now** |
| `email()` | `.email()` | String-specific format validation. | **Can Support Now** |
| `url()` | `.url()` | String-specific format validation. | **Can Support Now** |
| `uuid()` | `.uuid()` | String-specific format validation. | **Can Support Now** |
| `cuid()` | `.cuid()` | String-specific format validation. | **Can Support Now** (less common, maybe lower priority) |
| `datetime()` | `.datetime()` | String-specific format validation. | **Can Support Now** |
| `regex(RegExp re)` | `.regex(re)` | String-specific pattern matching. | **Can Support Now** |
| `startsWith(String s)` | `.startsWith(s)` | String-specific content validation. | **Can Support Now** |
| `endsWith(String s)` | `.endsWith(s)` | String-specific content validation. | **Can Support Now** |
| `trim()` | `.trim()` | **Transformation**. | **Cannot Support Now**. Our architecture validates, it doesn't transform the input value. |
| `toLowerCase()` | `.toLowerCase()` | **Transformation**. | **Cannot Support Now**. |
| `toUpperCase()` | `.toUpperCase()` | **Transformation**. | **Cannot Support Now**. |

---

### **3. `IntegerSchema` & `DoubleSchema` Extensions**

These will be in `lib/src/schemas/extensions/num_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `gt(num n)` / `greaterThan(num n)` | `.gt(n)` | Numeric comparison. | **Can Support Now** |
| `gte(num n)` / `min(num n)` | `.gte(n)` | Numeric comparison. | **Can Support Now** |
| `lt(num n)` / `lessThan(num n)` | `.lt(n)` | Numeric comparison. | **Can Support Now** |
| `lte(num n)` / `max(num n)` | `.lte(n)` | Numeric comparison. | **Can Support Now** |
| `isPositive()` | `.positive()` | Numeric sign check. | **Can Support Now** |
| `isNegative()` | `.negative()` | Numeric sign check. | **Can Support Now** |
| `isNonPositive()` | `.nonpositive()` | Numeric sign check. | **Can Support Now** |
| `isNonNegative()` | `.nonnegative()` | Numeric sign check. | **Can Support Now** |
| `multipleOf(num n)` / `step(num n)` | `.multipleOf(n)` | Numeric divisibility check. | **Can Support Now** |
| `finite()` | `.finite()` | Checks if a number is finite. | **Can Support Now** (More relevant for `DoubleSchema`) |
| `safe()` | `.safe()` | Checks if an integer is a "safe" integer. | **Can Support Now** (For `IntegerSchema`) |

---

### **4. `ListSchema` Extensions**

These will be in `lib/src/schemas/extensions/list_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `minLength(int n)` | `.min(n)` | List-specific length validation. | **Can Support Now** |
| `maxLength(int n)` | `.max(n)` | List-specific length validation. | **Can Support Now** |
| `length(int n)` | `.length(n)` | List-specific length validation. | **Can Support Now** |
| `nonempty()` | `.nonempty()` | A convenient shorthand for `.minLength(1)`. | **Can Support Now** |

---

### **5. `ObjectSchema` Extensions**

This is where things get more complex. Our `ObjectSchema` is powerful but doesn't have all of Zod's dynamic capabilities.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `pick(List<String> keys)` | `.pick()` | **Structural Transformation**. | **Cannot Support Now**. This creates a *new* schema type by altering the properties. Our `copyWith` can't do this. |
| `omit(List<String> keys)` | `.omit()` | **Structural Transformation**. | **Cannot Support Now**. Same reason as `pick`. |
| `partial()` | `.partial()` | **Structural Transformation**. | **Cannot Support Now**. This changes the `required` status of all properties, which would require deep schema cloning. |
| `deepPartial()` | `.deepPartial()` | **Structural Transformation**. | **Cannot Support Now**. Even more complex than `partial`. |
| `merge(ObjectSchema other)` | `.merge()` | Combines two object schemas. | **Can Support Now (with limitations)**. We can implement a version of this that merges properties and `required` fields. |
| `passthrough()` | `.`passthrough()` | Allows unknown keys. | **Already Supported**. This is our `allowAdditionalProperties: true` flag. We could add a fluent method for it. |
| `strict()` | `.strict()` | Disallows unknown keys. | **Already Supported**. This is our `allowAdditionalProperties: false` flag. We could add a fluent method for it. |
| `catchall(AckSchema valueSchema)` | `.catchall()` | Validates all unknown keys against a schema. | **Cannot Support Now**. Our current implementation only checks for the *presence* of additional properties, not their *values*. |

---

### **6. Features We Cannot Support (The "Why")**

*   **Transformations (`.trim()`, `.toLowerCase()`, `.transform()`, `.refine()`)**:
    *   **Reason:** Our architecture has a clear separation of concerns: a schema's job is to **validate** an input and return a `SchemaResult` with either the original (but correctly typed) value or an error. It does not **mutate or transform** the input value. Zod's `.transform()` and `.refine()` fundamentally change the output value, which is a different paradigm. Supporting this would require a major redesign of the `parseAndValidate` flow.

*   **Structural Transformations (`.pick()`, `.omit()`, `.partial()`)**:
    *   **Reason:** These methods in Zod dynamically create *new* schema definitions at runtime by introspecting and cloning existing schemas. Our `copyWith` pattern is great for creating a modified *copy*, but it's not designed for this level of dynamic, structural alteration (e.g., removing keys from the `properties` map and the `requiredProperties` list simultaneously). It's possible, but would require significantly more complex logic in the `ObjectSchema`'s `copyWith` method.

*   **Asynchronous Validation (`.refineAsync()`)**:
    *   **Reason:** Our entire validation pipeline (`validate` -> `parseAndValidate` -> `SchemaResult`) is synchronous. Introducing `async` validation would require making the entire pipeline return `Future<SchemaResult>`, which is a massive and breaking change across the entire library. 