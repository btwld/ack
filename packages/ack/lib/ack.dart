/// ACK - A Comprehensive Kotlin-inspired validation library for Dart
library;

// Constraints
export 'src/constraints/constraint.dart';
// Context
export 'src/context.dart';
export 'src/schemas/extensions/list_schema_extensions.dart';
export 'src/schemas/extensions/numeric_extensions.dart';
export 'src/schemas/extensions/object_schema_extensions.dart';
// Extensions
export 'src/schemas/extensions/string_schema_extensions.dart';
// Core schemas
export 'src/schemas/schema.dart';
export 'src/validation/ack_exception.dart';
export 'src/validation/schema_error.dart';
// Validation results
export 'src/validation/schema_result.dart';
