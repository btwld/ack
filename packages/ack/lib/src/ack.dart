import 'schemas/extensions/string_schema_extensions.dart';
import 'schemas/schema.dart';

/// The main entry point for creating schemas with the Ack validation library.
///
/// Provides a fluent API for creating various schema types.
final class Ack {
  /// Creates a string schema.
  static StringSchema string() => const StringSchema();

  /// Creates a literal string schema that only accepts the exact [value].
  /// Similar to Zod's `z.literal("value")`.
  static StringSchema literal(String value) => string().literal(value);

  /// Creates an integer schema.
  static IntegerSchema integer() => const IntegerSchema();

  /// Creates a double schema.
  static DoubleSchema double() => const DoubleSchema();

  /// Creates a boolean schema.
  static BooleanSchema boolean() => const BooleanSchema();

  /// Creates an object schema with the given properties.
  static ObjectSchema object(
    Map<String, AckSchema> properties, {
    List<String> required = const [],
    bool additionalProperties = false,
  }) =>
      ObjectSchema(
        properties.cast(),
        required: required,
        additionalProperties: additionalProperties,
      );

  /// Creates a discriminated object schema for polymorphic validation.
  static DiscriminatedObjectSchema discriminated({
    required String discriminatorKey,
    required Map<String, ObjectSchema> schemas,
  }) =>
      DiscriminatedObjectSchema(
        discriminatorKey: discriminatorKey,
        schemas: schemas,
      );

  /// Creates a list schema with the given item schema.
  static ListSchema<T> list<T extends Object>(AckSchema<T> itemSchema) =>
      ListSchema(itemSchema);

  /// Creates an enum schema for validating enum values.
  static EnumSchema<T> enumValues<T extends Enum>(List<T> values) =>
      EnumSchema(values: values);

  /// Creates a schema that can be one of many types.
  static AnyOfSchema anyOf(List<AckSchema> schemas) => AnyOfSchema(schemas);
}
