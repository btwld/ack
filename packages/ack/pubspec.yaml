name: ack
description: A simple validation library for Dart
version: 0.2.0
repository: https://github.com/btwld/ack

environment:
  sdk: ^3.3.0

dependencies:
  meta: ^1.15.0

dev_dependencies:
  coverage: ^1.11.1
  dart_code_metrics_presets: ^2.22.0
  lints: ^5.0.0
  test: ^1.24.0
