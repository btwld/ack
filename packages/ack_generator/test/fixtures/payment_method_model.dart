import 'package:ack/ack.dart';

part 'payment_method_model.g.dart';

/// A discriminated union representing different payment methods
/// This demonstrates the ideal developer experience for sealed classes with Ack
@Schema(
  description:
      'A payment method that can be credit card, bank transfer, or digital wallet',
  discriminatedKey: 'type',
)
sealed class PaymentMethod {
  /// The payment method type used for discrimination
  final String type;

  /// Optional payment amount limit
  final double? limit;

  /// Whether this payment method is the default
  final bool isDefault;

  const PaymentMethod({
    required this.type,
    this.limit,
    this.isDefault = false,
  });

  /// Convert to JSON representation
  Map<String, dynamic> toJson();
}

/// Credit card payment method
@Schema(
  description: 'Credit card payment with card details',
  discriminatedValue: 'credit_card',
)
class CreditCardPayment extends PaymentMethod {
  /// Card number (last 4 digits for security)
  @MinLength(4)
  final String cardNumber;

  /// Cardholder name
  @IsNotEmpty()
  final String cardholderName;

  /// Card expiry date
  final String expiryDate;

  /// Card brand (Visa, MasterCard, etc.)
  final String brand;

  const CreditCardPayment({
    super.limit,
    super.isDefault,
    required this.cardNumber,
    required this.cardholderName,
    required this.expiryDate,
    required this.brand,
  }) : super(type: 'credit_card');

  @override
  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'cardNumber': cardNumber,
      'cardholderName': cardholderName,
      'expiryDate': expiryDate,
      'brand': brand,
      if (limit != null) 'limit': limit,
      'isDefault': isDefault,
    };
  }
}

/// Bank transfer payment method
@Schema(
  description: 'Bank transfer payment with account details',
  discriminatedValue: 'bank_transfer',
)
class BankTransferPayment extends PaymentMethod {
  /// Bank account number
  @IsNotEmpty()
  final String accountNumber;

  /// Bank routing number
  @IsNotEmpty()
  final String routingNumber;

  /// Bank name
  @IsNotEmpty()
  final String bankName;

  /// Account holder name
  @IsNotEmpty()
  final String accountHolderName;

  const BankTransferPayment({
    super.limit,
    super.isDefault,
    required this.accountNumber,
    required this.routingNumber,
    required this.bankName,
    required this.accountHolderName,
  }) : super(type: 'bank_transfer');

  @override
  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'accountNumber': accountNumber,
      'routingNumber': routingNumber,
      'bankName': bankName,
      'accountHolderName': accountHolderName,
      if (limit != null) 'limit': limit,
      'isDefault': isDefault,
    };
  }
}

/// Digital wallet payment method
@Schema(
  description: 'Digital wallet payment like PayPal, Apple Pay, etc.',
  discriminatedValue: 'digital_wallet',
  additionalProperties: true,
  additionalPropertiesField: 'metadata',
)
class DigitalWalletPayment extends PaymentMethod {
  /// Wallet provider (PayPal, Apple Pay, Google Pay, etc.)
  @IsNotEmpty()
  final String provider;

  /// Wallet account identifier
  @IsEmail()
  final String walletId;

  /// Additional wallet-specific metadata
  final Map<String, dynamic> metadata;

  const DigitalWalletPayment({
    super.limit,
    super.isDefault,
    required this.provider,
    required this.walletId,
    this.metadata = const {},
  }) : super(type: 'digital_wallet');

  @override
  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'provider': provider,
      'walletId': walletId,
      ...metadata,
      if (limit != null) 'limit': limit,
      'isDefault': isDefault,
    };
  }
}
