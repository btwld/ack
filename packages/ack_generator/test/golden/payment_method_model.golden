// GENERATED CODE - DO NOT MODIFY BY HAND

// ignore_for_file: unnecessary_this, unnecessary_new, unnecessary_const, prefer_collection_literals
// ignore_for_file: lines_longer_than_80_chars, unnecessary_null_checks, non_constant_identifier_names

part of 'payment_method_model.dart';

// **************************************************************************
// AckSchemaGenerator
// **************************************************************************

/// Generated base schema for PaymentMethod with inheritance support
/// A payment method that can be credit card, bank transfer, or digital wallet
class PaymentMethodSchema extends BaseSchema<PaymentMethodSchema> {
  /// Default constructor for parser instances
  const PaymentMethodSchema();

  /// Private constructor for validated instances
  const PaymentMethodSchema._valid(Map<String, Object?> data)
      : super.valid(data);

  static final DiscriminatedObjectSchema schema = Ack.discriminated(
    discriminatorKey: 'type',
    schemas: {
      'credit_card': CreditCardPaymentSchema.schema,
      'bank_transfer': BankTransferPaymentSchema.schema,
      'digital_wallet': DigitalWalletPaymentSchema.schema,
    },
  );

  static final ObjectSchema baseSchema = Ack.object(
    {
      'type': Ack.string,
      'limit': Ack.double.nullable(),
      'isDefault': Ack.boolean,
    },
    required: ['type'],
    additionalProperties: false,
  );

  /// Parse with validation - core implementation
  @override
  PaymentMethodSchema parse(Object? data) {
    final result = definition.validate(data);
    if (result.isOk) {
      final validatedData = Map<String, Object?>.from(
        result.getOrThrow() as Map<String, Object?>,
      );
      return PaymentMethodSchema._valid(validatedData);
    }
    throw AckException(result.getError());
  }

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<PaymentMethodSchema>(
      (data) => PaymentMethodSchema(data),
    );
    CreditCardPaymentSchema.ensureInitialize();
    BankTransferPaymentSchema.ensureInitialize();
    DigitalWalletPaymentSchema.ensureInitialize();
  }

  @override
  ObjectSchema get definition => schema;

  String get type => getValue<String>('type')!;

  double? get limit => getValue<double>('limit');

  bool get isDefault => getValue<bool>('isDefault')!;

  R when<R>({
    required R Function(CreditCardPaymentSchema) creditCard,
    required R Function(BankTransferPaymentSchema) bankTransfer,
    required R Function(DigitalWalletPaymentSchema) digitalWallet,
  }) =>
      switch (type) {
        'credit_card' => creditCard(CreditCardPaymentSchema(toMap())),
        'bank_transfer' => bankTransfer(BankTransferPaymentSchema(toMap())),
        'digital_wallet' => digitalWallet(DigitalWalletPaymentSchema(toMap())),
        _ => throw StateError('Unknown paymentmethod type: $type'),
      };

  R maybeWhen<R>({
    R Function(CreditCardPaymentSchema)? creditCard,
    R Function(BankTransferPaymentSchema)? bankTransfer,
    R Function(DigitalWalletPaymentSchema)? digitalWallet,
    required R Function() orElse,
  }) =>
      switch (type) {
        'credit_card' =>
          creditCard?.call(CreditCardPaymentSchema(toMap())) ?? orElse(),
        'bank_transfer' =>
          bankTransfer?.call(BankTransferPaymentSchema(toMap())) ?? orElse(),
        'digital_wallet' =>
          digitalWallet?.call(DigitalWalletPaymentSchema(toMap())) ?? orElse(),
        _ => orElse(),
      };

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}

/// Generated schema for CreditCardPayment extending PaymentMethodSchema
/// Credit card payment with card details
class CreditCardPaymentSchema extends PaymentMethodSchema {
  /// Default constructor for parser instances
  const CreditCardPaymentSchema();

  /// Private constructor for validated instances
  const CreditCardPaymentSchema._valid(Map<String, Object?> data)
      : super.valid(data);

  static final ObjectSchema schema = PaymentMethodSchema.baseSchema.extend(
    {
      'cardNumber': Ack.string.minLength(4),
      'cardholderName': Ack.string.notEmpty(),
      'expiryDate': Ack.string,
      'brand': Ack.string
    },
    required: ['cardNumber', 'cardholderName', 'expiryDate', 'brand'],
    additionalProperties: false,
  );

  /// Parse with validation - core implementation
  @override
  CreditCardPaymentSchema parse(Object? data) {
    final result = definition.validate(data);
    if (result.isOk) {
      final validatedData = Map<String, Object?>.from(
        result.getOrThrow() as Map<String, Object?>,
      );
      return CreditCardPaymentSchema._valid(validatedData);
    }
    throw AckException(result.getError());
  }

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<CreditCardPaymentSchema>(
      (data) => CreditCardPaymentSchema(data),
    );
  }

  @override
  ObjectSchema get definition => schema;

  String get cardNumber => getValue<String>('cardNumber')!;

  String get cardholderName => getValue<String>('cardholderName')!;

  String get expiryDate => getValue<String>('expiryDate')!;

  String get brand => getValue<String>('brand')!;

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}

/// Generated schema for BankTransferPayment extending PaymentMethodSchema
/// Bank transfer payment with account details
class BankTransferPaymentSchema extends PaymentMethodSchema {
  /// Default constructor for parser instances
  const BankTransferPaymentSchema();

  /// Private constructor for validated instances
  const BankTransferPaymentSchema._valid(Map<String, Object?> data)
      : super.valid(data);

  static final ObjectSchema schema = PaymentMethodSchema.baseSchema.extend(
    {
      'accountNumber': Ack.string.notEmpty(),
      'routingNumber': Ack.string.notEmpty(),
      'bankName': Ack.string.notEmpty(),
      'accountHolderName': Ack.string.notEmpty()
    },
    required: [
      'accountNumber',
      'routingNumber',
      'bankName',
      'accountHolderName'
    ],
    additionalProperties: false,
  );

  /// Parse with validation - core implementation
  @override
  BankTransferPaymentSchema parse(Object? data) {
    final result = definition.validate(data);
    if (result.isOk) {
      final validatedData = Map<String, Object?>.from(
        result.getOrThrow() as Map<String, Object?>,
      );
      return BankTransferPaymentSchema._valid(validatedData);
    }
    throw AckException(result.getError());
  }

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<BankTransferPaymentSchema>(
      (data) => BankTransferPaymentSchema(data),
    );
  }

  @override
  ObjectSchema get definition => schema;

  String get accountNumber => getValue<String>('accountNumber')!;

  String get routingNumber => getValue<String>('routingNumber')!;

  String get bankName => getValue<String>('bankName')!;

  String get accountHolderName => getValue<String>('accountHolderName')!;

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}

/// Generated schema for DigitalWalletPayment extending PaymentMethodSchema
/// Digital wallet payment like PayPal, Apple Pay, etc.
class DigitalWalletPaymentSchema extends PaymentMethodSchema {
  /// Default constructor for parser instances
  const DigitalWalletPaymentSchema();

  /// Private constructor for validated instances
  const DigitalWalletPaymentSchema._valid(Map<String, Object?> data)
      : super.valid(data);

  static final ObjectSchema schema = PaymentMethodSchema.baseSchema.extend(
    {
      'provider': Ack.string.notEmpty(),
      'walletId': Ack.string.email(),
      'metadata': Ack.object({}, additionalProperties: true)
    },
    required: ['provider', 'walletId'],
    additionalProperties: false,
  );

  /// Parse with validation - core implementation
  @override
  DigitalWalletPaymentSchema parse(Object? data) {
    final result = definition.validate(data);
    if (result.isOk) {
      final validatedData = Map<String, Object?>.from(
        result.getOrThrow() as Map<String, Object?>,
      );
      return DigitalWalletPaymentSchema._valid(validatedData);
    }
    throw AckException(result.getError());
  }

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<DigitalWalletPaymentSchema>(
      (data) => DigitalWalletPaymentSchema(data),
    );
  }

  @override
  ObjectSchema get definition => schema;

  String get provider => getValue<String>('provider')!;

  String get walletId => getValue<String>('walletId')!;

  Map<String, dynamic> get metadata =>
      getValue<Map<String, Object?>>('metadata')!;

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}
